### PR TITLE
Add a performance testing app

### PR DESCRIPTION
**Pull request by igorT**
This is only the app part, Github action that builds it coming separately, as we need to merge this in two commits for the action to not fail immediately.

Biggest part of this PR is the new ember app, with the default blueprints.

### CODE REVIEW

**Review by rwjblue — CHANGES_REQUESTED** [review]
A few high level things:

* Using a more typical workspaces setup would make this easier to maintain over time (e.g. it would make only a single `yarn.lock` instead of two)
* Can you update the `CONTRIBUTING.md` with a reference to the overall goal / intent? Specifically, that the perf testing app exists and where to go find more info about it (that should link to the testing apps own README.md). Should also include some info about how we are expected to maintain the perf testing app over time (without a plan, this will just bit rot).
* Need to add basic application level tests for m3-perf-testing-app that is ran by the top level CI. There is an issue right now, that should have failed CI here, but didn't (missing `models/sample-data` module).

File: m3-perf-testing-app/app/routes/materializing.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import Route from '@ember/routing/route';
+import { action } from '@ember/object';
+import generateSampleData from '../models/sample-data';
```
[the remark this answers is not in the document]
Comment (igorT) [reply m3-perf-testing-app/app/routes/materializing.js:3]: Good catch, I managed to erase when cleaning up, splitting into two PRs. https://github.com/hjdivad/ember-m3/pull/1157 would have caught the failure, but will fix and add a separate test.

File: m3-perf-testing-app/app/routes/materializing.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+export default class Materializing extends Route {
+  model() {
+    performance.mark('start-loading');
+    for (let i = 0; i < 10000; i++) {
+      this.store.pushPayload(
+        'com.example.bookstore.search-results',
+        generateSampleData(i)
```
[the remark this answers is not in the document]
Comment (igorT) [reply m3-perf-testing-app/app/routes/materializing.js:11]: I wasn't very concerned by having it be inline, as it was going to be the same constant between `experiment` and `control` and is a hardcoded json blob so wasn't going to impact the perf difference, but good point I'll take it out.

File: .npmignore [hunk 3]
```diff
@@ -0,0 +1,33 @@
+# See https://help.github.com/ignore-files/ for more about ignoring files.
```
[the remark this answers is not in the document]
Comment (igorT) [reply .npmignore:1]: Ah good catch didn't realize thats where we defined them, thought we were piggybacking off of .gitignore

[the remark this answers is not in the document]
Comment (igorT) [reply .npmignore:1]: Removed

File: m3-perf-testing-app/config/ember-cli-update.json [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    {
+      "name": "ember-cli",
+      "version": "3.26.1",
+      "blueprints": [
+        {
+          "name": "app",
+          "outputRepo": "https://github.com/ember-cli/ember-new-output",
+          "codemodsSource": "ember-app-codemods-manifest@1",
+          "isBaseBlueprint": true,
+          "options": []
```
[the remark this answers is not in the document]
Comment (igorT) [reply m3-perf-testing-app/config/ember-cli-update.json:13]: done

File: m3-perf-testing-app/app/routes/materializing.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (igorT) [reply m3-perf-testing-app/app/routes/materializing.js:11]: done

File: m3-perf-testing-app/app/routes/materializing.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (igorT) [reply m3-perf-testing-app/app/routes/materializing.js:3]: fixed

File: m3-perf-testing-app/README.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# m3-perf-testing-app
+
+This app is used for tracking performance improvements and regressions, using the perf-check workflow
```
[the remark this answers is not in the document]
Comment (igorT) [reply m3-perf-testing-app/README.md:3]: I have added more detail in this README, let me know if you think there is still a need for a per route Readme

**Comment by igorT** [comment]
> * Using a more typical workspaces setup would make this easier to maintain over time
I have personally found that the workspaces setup has a decent overhead, and papercuts over time so given that we only have one extra package here, seemed nicer and easier to go with the subfolder solution.

Addressed other high level comments

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -65,6 +65,20 @@ jobs:
...
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-node@v1
+        with:
+          node-version: '12.x' # min node version
+      - name: yarn install
+        run: cd m3-perf-testing-app && yarn --frozen-lockfile --install
+      - name: test:performance-app
+        run: cd m3-perf-testing-app && yarn test
```
Comment (rwjblue) [inline .github/workflows/ci.yml:80]: FWIW, you don't have to `cd ... && other-thing` like this, you can tell GH actions to use whatever working dir you want:

```suggestion
  test-performance-app:
    name: performance-app
    needs: [lint]
    runs-on: ubuntu-latest
    
    # run all of the commands below inside the perf test app
    defaults:
      run:
        working-directory: m3-perf-testing-app

    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v1
        with:
          node-version: '12.x' # min node version
      - name: yarn install
        run: yarn --frozen-lockfile --install
      - name: test:performance-app
        run: yarn test 
```

Alternatively you could specify the working directory for each `run:` command like this:

```js
  test-performance-app:
    name: performance-app
    needs: [lint]
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v1
        with:
          node-version: '12.x' # min node version
      - name: yarn install
        run: yarn --frozen-lockfile --install
        working-directory: m3-perf-testing-app
      - name: test:performance-app
        run: yarn test 
        working-directory: m3-perf-testing-app
```

Comment (rwjblue) [reply .github/workflows/ci.yml:80]: Some docs on `defaults.run` here: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_iddefaults

File: m3-perf-testing-app/README.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# m3-perf-testing-app
+
+This app is used for tracking performance improvements and regressions, using the perf-check CI workflow
+
+It currently has two routes used for tracking performance `materializing` and `rendering`.
+
+Materializing tests pushing a sample m3 payload which includes includes plain attributes, managed arrays, embedded models and references into the store,
+and accessing a few of their properties, in order to test overall push speed and the speed of going through the attribute resolution.
+
+Renderirng tests pushing and rendering a sample m3 payload which includes plain attributes, managed arrays, embedded models and references, in order to test the core non network workflow of typical m3 usage.
```
Comment (rwjblue) [inline m3-perf-testing-app/README.md:10]:
```suggestion
It currently has two routes used for tracking performance `materializing` and `rendering`.

## Materializing

The materializing route tests pushing a sample m3 payload which includes plain attributes, managed arrays, embedded models and references into the store,
and accessing a few of their properties.

The goal of this test is to cover the overall push speed and the speed of going through the attribute resolution.

## Rendering

Rendering tests pushing and rendering a sample m3 payload which includes plain attributes, managed arrays, embedded models and references.

The goal of this test is to cover the non-networking workflow of typical m3 usage.
```

File: m3-perf-testing-app/app/serializers/application.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import { assign } from '@ember/polyfills';
+
+export default class ApplicationSerializer {
+  constructor(createArgs) {
+    assign(this, createArgs);
+  }
```
Comment (rwjblue) [inline m3-perf-testing-app/app/serializers/application.js:6]: We might not need to use `@ember/polyfills` for this. I think we can reasonably decide not to support IE11 in our test app.

```suggestion
export default class ApplicationSerializer {
  constructor(createArgs) {
    Object.assign(this, createArgs);
  }
```

File: .github/workflows/ci.yml (hunk 6 above)
Comment (igorT) [reply .github/workflows/ci.yml:80]: Oh neat thanks!

**Review by rwjblue — APPROVED** [review]

**Pull request merged**